openedx/core/djangoapps/discussions/docs/decisions/0004-in-context-discussions-linking.rst

```diff
@@ -0,0 +1,137 @@
+
+This is OK for the existing setup because this is just a caching mechanism and
+the source of truth for this mapping is the XBlock itself, which stores the
+discussion id. On course publish this information is cached to
+`DiscussionsIdMapping`.
+
+For the new discussions system though, this mapping would be the source of
+truth for the link between discussions and units, so we should use a model
+where each such link is encoded as a row in the database.
+
+Decision
+--------
+
+Since the discussions settings need to be stored in the course structure we
+should create a new JSON structure in the course that matches the structure
+of `plugin_settings`. This can then be used to store not just the settings
+for the inbuilt discussions provider, but for any discussions provider in the
+future.
+
+When a course is published, we can copy over all the `plugin_settings` to the
+course in a JSON field called `dicussions_settings` with the following
+structure:
+
+.. code-block:: JSON
+
+    {
+        "discussions_enable_in_context": bool,
+        "discussions_enable_graded_units": bool,
+        "discussions_custom_visibility": bool,
+        "edx-next": {
+            "discussions_group_at_subsection": bool,
+        }
+    }
+
+The `edx-next` key here represents the provider id, allowing for potentially
+multiple provider configs to coexist in case of switching providers etc.
+Settings outside this key are those that are applicable to all providers. Note
+that they may not be supported by all providers though, in which case they will
+simply be ignored.
+
+To store Unit-level discussions settings, we can simply add a boolean field
+to the Unit block that specifies whether it is discussable or not. To be
+consistent with the above names we can call this field `discussions_enabled`.
+
+A signal can be created using the new Hooks extension system proposed in OEP-50
+that is triggered when discussions settings change. This signal can encapsulate
+all the data needed for setting up discussions from the modulestore. It can
+traverse through all teh Units in the course that match the criterion from the
+dicussions settings and provide the needed details as part of the signal data.
+
+A handler for the above signal, we create the discussion topics in
+`cs_comments_service` and add a mapping. If an existing unit with discussions
+is removed, we will disable the link but not delete the data.
+
+The discussion grouping at subsections will simply combine the topics from all
+the units in the subsection and provide a unified view across the subsection.
+This setting will mainly be ignored and will likely only be used by the APIs
+or potentially the frontend directly.
+
+The mapping between discussion ids and units is also a very simple model:
+
+.. code-block:: python
+
+    class DiscussionTopicLink:
+        course_key: CourseKey
+        usage_key: UsageKey
+        title: str
+        group_id: int
+        discussion_provider_id: str
```

Comment:
Why is this a `str` instead of a foreign key to a model representing the discussion provider?

Reply:
Discussion providers don't currently have a model representation. They are currently all in a dict here: 

https://github.com/edx/edx-platform/blob/84441bfd3a28862ab01834accb0c41a49c57a924/openedx/core/djangoapps/discussions/models.py#L90-L221